upstairs/src/lib.rs

```diff
@@ -1291,22 +1291,28 @@ where
                         if negotiated != 4 {
                             bail!("Received ExtentVersions out of order!");
                         }
+                        let active = up.active.lock().await;
+                        let up_state = active.up_state;
+                        let mut ds = up.downstairs.lock().await;
+                        drop(active);
```

Comment:
Is it required that the `up.active` lock be taken before the `up.downstairs` lock?, or that `active` be dropped before taking the `up.downstairs` lock?

Reply:
Yes, it is required that we have the upstairs state when we change the downstairs state.
To make sure we don't deadlock, we always get the upstairs state first.
 We need to know what the current upstairs state is (active or not) when we transition
downstairs states to know if that transition is valid.

We must get the downstairs lock while holding the upstairs active lock so we know we are
comparing the states of both at the "same time", and we don't release the downstairs lock
until we have made (or not made) any transition for the downstairs.